### PR TITLE
feat: add `TableBuilder` API for passing options

### PR DESCRIPTION
**Pull request by kazdy**
## Description

Add `TableBuilder` to handle Table construction.

Allow to explicitly set hudi_options and storage_options via builder in rust or via python bindings when creating new Table instance.

```rust
let hudi_table = TableBuilder::from_base_uri("/tmp/hudi_table")
  .with_hudi_options([("hoodie.read.as.of.timestamp", "0")])
  .with_storage_options([("AWS_REGION", "us-east-1")])
  .build()
```

similarly for python

```python
hudi_table = HudiTableBuilder.from_base_uri(
  "/tmp/hudi_table"
  ).with_hudi_options({
    "hoodie.read.as.of.timestamp": "0"
  }).with_storage_options({
    "AWS_REGION": "us-east-1"
  }).build()
```

resolves: #139 

## How are the changes test-covered

- [ ] N/A
- [x] Automated tests (unit and/or integration tests)
- [ ] Manual tests
  - [ ] Details are described below

### CODE REVIEW

**Comment by kazdy** [comment]
@xushiyan please let me know if this is a direction you're fine with

File: crates/core/src/table/mod.rs [hunk 1]
```diff
@@ -180,91 +169,6 @@ impl Table {
             .register_object_store(runtime_env.clone());
     }
 
-    async fn load_configs<I, K, V>(
```
Comment (xushiyan) [inline crates/core/src/table/mod.rs:183]: some of these util functions were updated in main

File: crates/core/src/table/builder.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    }
+    pub fn with_hudi_options(mut self, hudi_options: HashMap<String, String>) -> Self {
+        self.hudi_options = Some(hudi_options);
+        self
+    }
+
+    pub fn with_storage_options(mut self, storage_options: HashMap<String, String>) -> Self {
+        self.storage_options = Some(storage_options);
+        self
+    }
```
Comment (xushiyan) [inline crates/core/src/table/builder.rs:57]: a builder pattern is helpful for users. can you design such that it takes in generic options? we don't have to distinguish hudi option vs storage option from the user side, the build() function should handle it internally.

**Review by xushiyan — COMMENTED** [review]
there are some refactoring around this. can you check out https://github.com/apache/hudi-rs/pull/161 first, and adjust this PR accordingly please?

File: python/hudi/_internal.pyi [hunk 3]
```diff
@@ -38,7 +38,8 @@ class HudiTable:
     def __init__(
         self,
         table_uri: str,
-        options: Optional[Dict[str, str]] = None,
+        hudi_options: Optional[Dict[str, str]] = None,
+        storage_options: Optional[Dict[str, str]] = None,
```
Comment (xushiyan) [inline python/hudi/_internal.pyi:42]: you can revert this change - we keep python and rust api as close as possible

File: python/src/internal.rs [hunk 4]
```diff
@@ -101,12 +102,18 @@ pub struct HudiTable {
...
-    fn new(table_uri: &str, options: Option<HashMap<String, String>>) -> PyResult<Self> {
-        let _table = rt().block_on(Table::new_with_options(
-            table_uri,
-            options.unwrap_or_default(),
-        ))?;
+    #[pyo3(signature = (table_uri, hudi_options = None, storage_options = None))]
+    fn new(
+        table_uri: &str,
+        hudi_options: Option<HashMap<String, String>>,
+        storage_options: Option<HashMap<String, String>>,
```
Comment (xushiyan) [inline python/src/internal.rs:109]: this can be reverted as well

File: python/tests/test_table_read.py [hunk 5]
```diff
@@ -133,7 +133,9 @@ def test_sample_table(get_sample_table):
         },
     ]
 
-    table = HudiTable(table_path, {"hoodie.read.as.of.timestamp": "20240402123035233"})
+    table = HudiTable(
+        table_path, hudi_options={"hoodie.read.as.of.timestamp": "20240402123035233"}
+    )
```
Comment (xushiyan) [inline python/tests/test_table_read.py:138]: ditto

**Comment by xushiyan** [comment]
@kazdy i'm cutting release branch within 1 day. do you think you can update this soon?

File: crates/core/src/table/builder.rs (hunk 2 above)
Comment (kazdy) [reply crates/core/src/table/builder.rs:57]: Ok I'll do it, but I'd like to still give users (and myself as a user as well) option to be specific.
I wanted to have storage_options because from the first look I did not figure out options handle both hudi and storage opts.

So it'll be 
TableBuilder::with_options() -> handle generic options
TableBuilder::with_hudi_options()
TableBuilder::with_storage_options()

I hope that's fine

**Comment by kazdy** [comment]
> @kazdy i'm cutting release branch within 1 day. do you think you can update this soon?

yes, I'll give it a try

File: python/hudi/_internal.pyi (hunk 3 above)
Comment (kazdy) [reply python/hudi/_internal.pyi:42]: will do, I'll introduce HudiTableBuilder in python bindings instead

File: crates/core/src/table/builder.rs (hunk 2 above)
Comment (xushiyan) [reply crates/core/src/table/builder.rs:84]: @kazdy can you please clarify the benefit of using 

```
TableBuilder::with_hudi_options()
TableBuilder::with_storage_options()
```

when there is a generic api to take options and resolve internally, why would you still want to use an explicit api? using a generic api is one less thing caller needs to worry about. From API design pov, it'll be redundant unless being explicit gives additional logic behind it.

Comment (xushiyan) [reply crates/core/src/table/builder.rs:84]: if you add additional checks, like for hudi_options, do some validation (there is a util function for this) for the input, and for storage options, check they have certain prefixes, then that can be helpful in preventing mistakes

Comment (xushiyan) [reply crates/core/src/table/builder.rs:84]: but still, builder pattern usually does validation in the end when call build(), we don't want to panic during the building phase of with_xxx(). One point i can think of is, making input taken from explicit APIs have higher precedence than those taken by generic ones. This should be handled in build()

Comment (kazdy) [reply crates/core/src/table/builder.rs:84]: I think TableBuilder can give more control to the user and be more specific. I'm fine with having a generic API for ease of use, but I still think having this explicit distinction between hudi and storage options is nice and easy to understand when using the API. I was a bit confused at first and wondered if/where can I pass storage options.

Still Table has ::new_with_options() which is generic so that TableBuilder can be more specific?

I agree about validation being done in ::build().
I'm not sure if I'll be able to add more validation now, but I can give it a try in the next PR?

Comment (xushiyan) [reply crates/core/src/table/builder.rs:80]: > I'm not sure if I'll be able to add more validation now, but I can give it a try in the next PR?

yup sounds good

**Comment by xushiyan** [comment]
@kazdy I'll push some fixes and then get this landed for release

File: python/tests/test_table_builder.py
Comment (kazdy) [inline python/tests/test_table_builder.py:1]: @xushiyan please take a look

File: python/tests/test_table_read.py
Comment (kazdy) [inline python/tests/test_table_read.py:1]: @xushiyan please take a look, I took the freedom to split this one big test to a few smaller ones, I hope that's fine with you

File: crates/core/src/table/builder.rs [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+
+    async fn imbue_table_properties(
+        options: &mut HashMap<String, String>,
+        storage: Arc<Storage>,
+    ) -> anyhow::Result<()> {
+        let bytes = storage.get_file_data(".hoodie/hoodie.properties").await?;
+        let table_properties = parse_data_for_options(&bytes, "=")?;
+
+        // TODO: handle the case where the same key is present in both table properties and options
```
Comment (kazdy) [inline crates/core/src/table/builder.rs:161]: in such case, which option should take precedence?
@xushiyan

File: python/tests/test_table_read.py
Comment (xushiyan) [reply python/tests/test_table_read.py:1]: yup good improvement

File: crates/core/src/table/builder.rs (hunk 6 above)
Comment (xushiyan) [reply crates/core/src/table/builder.rs:167]: it's a bit complicated, there are difference cases pertaining to individual config - i'll fix this todo

**Review by xushiyan — APPROVED** [review]
